beamline.rst

```diff
@@ -0,0 +1,195 @@
+Beamline API
+~~~~~~~~~~~~
+
+Functionality that incorporates several different instruments or that does
+not necessarily belong to one particular instrument.
+
+
+
+**API Functions**
+
+.. code:: python
+
+   class CmdTuple(NamedTuple):
+    """
+    Describes a "generic" beamline procedure, where:
+
+    cmd_id: the string identifying the command, i.e. 'QUICK_REALIGN'
+    running: True if the procedure is running otherwise False
+    args: arguments to pass to the procedure
+    display_name: the name to display in the user interface i.e. "Quick Realign"
+    tool_tip: the tool tip
+    messages: any initial messages to display
+    """
+        cmd_id: str
+        state: CmdState
+        args: tuple
+        display_name: str
+        tool_tip: str
+        messages: list
+
+
+    def get_procedures() -> List[CmdTuple]:
+    """
+    :returns: a list of tuples each describing a beamline procedure,
+              such as: realignment, anneal
+
+    :rtype: List[CmdTuple]
+    """
+        pass
+
+        
+    def run_procedure(cmd_id:str, args) -> bool:
+    """
+    Runs procedure identified by cmd_id passes args as its arguments
+    
+    (Errors and progress that occurs is passed asynchronously via the
+     available signaling mechanism.)
+    """
+        
+    
+    def stop_procedure(cmd_id:str) -> bool:
+    """
+    Stops a running procedure identified by cmd_id
+
+    (Signal emitted if stopped or on timeout waiting for procedure to stop)
+    """
+
+
+    class Actuator(NamedTuple):
+    """
+    Describes an actuator commonly a motor, that is considered to be
+    "globally" available on the beamline. Such actuators could be shutters
+    of different kinds, i.e: fast shutter, safety shutter, other in-out
+    devices; beam stop, detector cover or even motors such as omega.
+
+    name: A unique name that identifies this actuator
+    value: The current position
+    msg: A message string, explaining state or value
+    state: The state of the actuator defined by ActuatorState
+           (MOVING, READY, ERROR ...)
+    """
+
+        name: str
+        value: float
+        msg: str
+        state: ActuatorState
+
+
+
+    def get_actuators() -> Dict[str, Actuator]:
+    """
+    :returns: A dictionary with all available actuators where the key
+              is the actuator name and the value the Actuator tuple
+    :rtype dict:            
+    """
+        pass
+
+
+    def get_actuator(name) -> Actuator:
+    """
+    :returns: The Actuator object identified by the given name
+    :rtype: Actuator
+    """
+       pass
+
+
+    def set_actuator(name, value) -> bool:
+    """
+    Tries to move the actuator identified by name to value
```

Comment:
Should not this be

def move_actuator(name:str, value:Any) -> bool:

or maybe

def set_actuator_value(...

Is is not confusing if get_actuator returns an Actuator object (NamedTuple), but set_actuator sets a value (probably float), rather than an Actuator object.

Reply:
Ah, yes it should be:

    set_actuator_value(name:str, value:float)

Its allright with me if everyone prefers move instead of set. I don't see any good reason for why we would need anything other than float (its basically position). Id prefer if we have a clear reason for the Any case ?

We only provide a value because that is the only thing that we can actually set, the rest of the data is un-interesting at this stage, so its just unnecessary to pass it around.

Reply:
I quite agree that we only need to set the value, it is a matter of naming/nitpicking, not of expanding the scope. 

For now I guess all we need is float, but it would depend on how widely this is used. 

There are basically two/three kinds of movables I can see:
Motors (with a continuous float value),
State switches (Beamstop.IN / Beamstop.OUT, but other objects could have more states)
'State motors' (?), like apertures, which are sort of a mixtyure, where you have a configurable list of floats as legal values. 

Would we want to use different mechanisms, or is it better to use the same abstract function for all  of them?

Another question in the same line is if we want to have upper and lower limits, and/or lists of allowed values.

Reply:
I see, well I considered In/Out to be 1/0. I have so far not seen aperture as an actuator in this way since its not something we would like to actually access directly from the UI.

 - The aperture would be moved indirectly through set_beam_size

 - Today we have positional motors, and IN/OUT devices that needs to be handled through this API call.

But why not extend the API to also handle "State Motors"

 - Yes we need limit values, it should be part of the Actuator Tuple